# Comparative-Genomics-of-Gloeobacter-and-Theobroma
Code and limited results repository for the assignment for the course ***Comparative genomics***, *MSc of Bioinformatics, KU Leuven, 2023*

**Contents:**
- BLAST both organisms against each other (`blast_both_all.pds`)
- Find BBHs for ortholog, paralog and co-ortholog detection (`bbh.py`)
- BLAST a *Gloeobacter* proteins against 25 handpicked proteomes to find all homologs (`get_homologs.pbs`)
- Parse results and identify homologs (`homologs.py`)
- Create a simple NJ phylogenetic tree using ClustalW (`homologs.phb`)
- Assess conservation across regions (`conservation.py`)
